standards/core/rln-contract.md

```diff
@@ -0,0 +1,301 @@
+
+### Will my deposit be slashed if I exceed the rate limit?
+
+This specification does not involve slashing.
+The aim of the deposit initially is to protect the network from denial-of-service attacks with bandwidth capping.
+
+### Do I need an extra deposit to extend a membership?
+
+Membership extension requires no additional deposit.
+The opportunity cost of locked-up capital plus gas fees for extension transactions make extensions non-free, which is sufficient for the initial mainnet deployment.
+
+### Why this particular epoch length?
+
+Epoch length is a global parameter set in the smart contract.
+Rate limits are defined in terms of the maximum allowed messages per epoch.
+There is a trade-off between short and long epochs.
+We chose an epoch length of `10` minutes as a reasonable middle-ground.
+
+On the one hand, longer epochs allow for better accommodating short-term usage peaks.
+On the other hand, long epochs increases memory requirements for RLN Relay nodes.
+
+Each message contains a nullifier that proves its validity in terms of RLN.
+Each RLN Relay node must keep in memory a nullifier log for the current epoch.
+Each nullifier plus metadata is `128` bytes (per message).
+With a `10`-minute epoch, one high-tier user with a `1` message per second rate limit generates up to `600 * 128 / 1024 = 75 KiB` of nullifier log data per epoch.
+This corresponds to:
+- for 1000 users: approximately `73 MiB`;
+- for 10 thousand users: approximately `732 MiB`.
+
+### Why is there a cap on the total rate limit?
+
+Total network bandwidth is a limited resource.
+We want to cap the total rate limit, at least in the initial mainnet deployment, to avoid overstretching the network's capabilities.
+
+### Why is there a minimal rate limit?
+
+The minimal rate limit prevents an attack where someone registers a large number of memberships with a tiny rate limit each, causing the RLN tree to contain too many elements.
+
+### Are there bulk discounts for high-rate memberships?
+
+For the initial mainnet deployment, there are no bulk discounts.
+Membership price is linearly proportional to its rate limit.
+We choose this pricing scheme for simplicity.
+Finding a pricing scheme with the right trade-off remains subject for future work, as
+high-rate memberships are arguably more efficient but can incentivize centralization.
+
+### Why only accept DAI?
+
+When choosing a token to accept, we considered the following criteria:
+- a stablecoin, as USD-denominated pricing is familiar for users and requires no oracle;
+- popular, high liquidity;
+- preferably decentralized;
+- with a reasonably good anti-censorship track record.
+
+Based on these criteria, we chose DAI for the initial mainnet deployment.
+Other tokens may be added in the future.
+
+## Security / Privacy Considerations
+
+Issuing membership-specific transactions (e.g., membership extension and deposit withdrawal) publicly links it to an Ethereum address.
```

Comment:
I Wonder if this should be in RLN Relay and if we should just point to RLN Relay's security section from here.  Just a thought.

Reply:
IMO, it can be either way.

The RLN Relay spec describes purely the P2P protocol. This spec describes the smart contract. Privacy issues (at least those that came to my mind, outlined here) originate at the intersection of both layers. Ideally, we should think of security and privacy holistically, taking all protocol components into consideration.